File: 0039-combination-sum/0039-combination-sum.java
class Solution {
    public List<List<Integer>> combinationSum(int[] candidates, int target) {
        List<List<Integer>> anw = new ArrayList<>();
        List<Integer> list = new ArrayList<>();
        helper(0, 0, target, candidates, anw, list);
        return anw;
    }

    private void helper(int start, 
    int sum,
    int target, 
    int[] candidates, 
    List<List<Integer>> anw, 
    List<Integer> list) {
        if (sum == target) {
            List<Integer> copy = new ArrayList<>(list);
            anw.add(copy);
            return;
        }
        if (start >= candidates.length || sum > target) return;
        list.add(candidates[start]);
        helper(start, sum + candidates[start], target, candidates, anw, list);
        list.remove(list.size() - 1);
        helper(start + 1, sum, target, candidates, anw, list);
    }
}

File: 0039-combination-sum/0039-combination-sum.cpp
class Solution {
void helper(vector<vector<int>>& res, vector<int>& candidates, vector<int>& curr, int i, int sum, int target){
    if (sum > target) return;
    if (sum == target) {
        res.push_back(curr);
        return;
    }

    if (i == candidates.size()) return;

    curr.push_back(candidates[i]);
    helper(res, candidates, curr, i, sum + candidates[i], target);
    curr.pop_back();
    helper(res, candidates, curr, i + 1, sum, target);
}
public:
    vector<vector<int>> combinationSum(vector<int>& candidates, int target) {
        vector<vector<int>> res;
        vector<int> curr;
        helper(res, candidates, curr, 0, 0, target);
        return res;
    }
};